### PR TITLE
Fix LightningCLI docs after overhaul of the documentation

### LINKED ISSUE #13249: Suggestions to improve LightningCLI documentation
## 📚 Documentation

With the recent complete overhaul of the documentation where things were split into expertise levels, some parts now seem misplaced or even incorrect. Here are a few suggestions to improve the documentation related to LightningCLI. 

- The first issue that I see is that there is no introduction. Why should someone use LightningCLI? There is too much emphasis on "reducing config boilerplate" and LightningCLI provides much more than reduction of boilerplate. Furthermore, it is perfectly possible to do everything in python files, so why bother with config files and CLIs? LightningCLI is about hyperparameters, reproducibility and best practices (e.g. separate code from configuration).

- Currently in master, before the LightningCLI section there is one on hyperparameters that describes how to implement CLIs with argparse. Anyone new to lightning will see this first and most likely use that. Only a long time later would realize that the same could be done with LightningCLI and requiring much less effort. LightningCLI should be introduced where the hyperparameters is first mentioned. And the argparse documentation be renamed to something like legacy CLI and recommend to use LightningCLI instead.

- In the old version of the documentation, the automatic save of the config is described at the beginning. Now it is barely mentioned. This is a central feature and one of the reasons why people should use LightningCLI for experiments.

- For the new structure of documentation, there was a bit of careless renaming and copy pasting. For example the section "Reproducibility with the LightningCLI" is not really something I would expect in a FAQ. And also the last two paragraphs where pasted from elsewhere which are not related to reproducibility. Another example is that the section "Argument linking" was renamed to "Connect two config files". Using a different terminology would be fine. But the new title is very misleading. That feature is not about connecting two config files.

### PR DESCRIPTION
**Pull request by mauvilsa**
## What does this PR do?

Fixes #13249

I found many more issues not mentioned in #13249. And still some things might need improvement, hopefully we can fix them with the reviews.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca

### CODE REVIEW

File: docs/source-pytorch/cli/lightning_cli.rst [hunk 1]
```diff
@@ -2,9 +2,25 @@
 
 .. _lightning-cli:
 
-############################
-Eliminate config boilerplate
```
Comment (awaelchli) [inline docs/source-pytorch/cli/lightning_cli.rst:6]: I preferred the previous title. 
To me, "Eliminate boilerplate" sounds more inviting to someone to try this feature.

Comment (awaelchli) [inline docs/source-pytorch/cli/lightning_cli.rst:6]: I preferred the previous title. 
To me, "Eliminate boilerplate" sounds more inviting to someone to try this feature.

File: docs/source-pytorch/cli/lightning_cli_advanced.rst [hunk 2]
```diff
@@ -1,113 +1,169 @@
...
+
+    If you are unfamiliar with yaml, have a look at the short introduction at `realpython.com#yaml-syntax
+    <https://realpython.com/python-yaml/#yaml-syntax>`__.
+
 
 ----
 
-***************************
-What is a yaml config file?
-***************************
```
Comment (awaelchli) [inline docs/source-pytorch/cli/lightning_cli_advanced.rst:14]: It was added by @williamfalcon, I would keep the example and combine it with your reference to the other tutorial.

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst [hunk 3]
```diff
@@ -209,69 +213,15 @@ If the CLI is implemented as :code:`LightningCLI(MyMainModel)` the configuration
         init_args:
           ...
 
-It is also possible to combine :code:`subclass_mode_model=True` and submodules, thereby having two levels of
-:code:`class_path`.
-
-
-Class type defaults
-^^^^^^^^^^^^^^^^^^^
```
Comment (awaelchli) [inline docs/source-pytorch/cli/lightning_cli_advanced_3.rst:217]: I couldn't find where this was moved or why it must be removed. Could you guide the reviewer here :)

**Review by awaelchli — APPROVED** [review]
Great improvements!

One suggestion for future work: In large PRs like this one, it would be great if you could give the reviewer some guidance on what has been moved where, as it is not so easy to discover just from the diff.

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst (hunk 3 above)
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_advanced_3.rst:217]: I did think about adding comments to guide, but then I forgot. Sorry!

This section was duplicated by mistake, see [lightning_cli_expert.rst#class-type-defaults](https://github.com/Lightning-AI/lightning/blob/master/docs/source-pytorch/cli/lightning_cli_expert.rst#class-type-defaults)

File: docs/source-pytorch/cli/lightning_cli.rst [hunk 4]
```diff
@@ -60,34 +76,52 @@ Advanced use
...
    :col_css: col-md-6
    :button_link: lightning_cli_faq.html
    :height: 150
 
+.. displayitem::
+   :header: Legacy CLIs
+   :description: Documentation for the legacy argparse-based CLIs
+   :col_css: col-md-6
+   :button_link: ../common/hyperparameters.html
+   :height: 150
```
[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source-pytorch/cli/lightning_cli.rst:123]: @awaelchli The CLI was recently promoted out of utilities into `pytorch_lightning.cli`. There's no advantage in the old functionality compared to the CLI. If legacy does not fit right with you (somewhat true since we haven't deprecated it), then maybe "old" is a better term.

**Review by carmocca — APPROVED** [review]

File: docs/source-pytorch/cli/lightning_cli.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli.rst:123]: "legacy" was just a suggestion. Could be some other word or just "argparse-based CLIs". My main goal was not to deprecate or demote it. It is just so that people new to lightning discover first `LightningCLI` and avoid them implementing things that wouldn't be required if they use `LightningCLI`.

File: docs/source-pytorch/cli/lightning_cli.rst (hunk 1 above)
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli.rst:6]: Similar to my previous comment. My guess is that people new to lightning will probably first search for hyperparameters and not reduction of boilerpate. In my opinion most important than the title was to combine the previous "Configure hyperparameters from the CLI" section with the one for LightningCLI so that when people go to it, there is clarity that they are both about implementing CLIs.

File: docs/source-pytorch/cli/lightning_cli_advanced_2.rst [hunk 5]
```diff
@@ -122,39 +111,37 @@ The Lightning CLI can convert every possible CLI flag into an environment variab
 
 .. code:: python
 
-    LightningCLI(env_parse=True)
+    cfg = LightningCLI(DemoModel, run=False, env_parse=True)
```
[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_advanced_2.rst:114]: Mistake.

File: docs/source-pytorch/cli/lightning_cli.rst (hunk 1 above)
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli.rst:6]: I think the title here should be the same as how it is shown on the main menu. For subsections could be other titles. Would be good to hear more opinions about titles.

File: docs/source-pytorch/cli/lightning_cli.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply docs/source-pytorch/cli/lightning_cli.rst:123]: @carmocca I am just surprised that we are calling it legacy and old, despite Python having the argparse library built-in and it's not going to disappear. No matter how awesome jsonargparse and LightningCLI are right now, there will be people who want to migrate to Lightning and just take their existing argparse code with them. They want to parse them and pass them into the model manually. To call this legacy or to even deprecate it (how even?) is a bit strong IMO. 

I suggest to recommend in the docs, in the hyperparameters.rst for example, whenever we can, to use the LightningCLI. This will promote the LightningCLI and still make it clear that you can use whatever.

[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source-pytorch/cli/lightning_cli.rst:123]: `argparse` itself is not legacy/old, but our `from_argparse_args` logic is as it has many limitations that the CLI does not. We will always allow the user to use whatever, but we are no longer investing resources into the pure argparse functionality ([example](https://github.com/Lightning-AI/lightning/issues/12974))

File: docs/source-pytorch/cli/lightning_cli_advanced_2.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_advanced_2.rst]: Fixed

File: docs/source-pytorch/cli/lightning_cli_advanced.rst (hunk 2 above)
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli_advanced.rst:14]: Done.

File: docs/source-pytorch/cli/lightning_cli.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli.rst:123]: Any suggestions on what I should do? Having both the `LightningCLI` and the `from_argparse_args` documentations under a single CLIs section is okay? If so, what should be the main title and the titles of the subsections?

**Comment by Borda** [comment]
@RobertLaurella could you pls check this out :otter:

File: docs/source-pytorch/cli/lightning_cli.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply docs/source-pytorch/cli/lightning_cli.rst:123]: Feel free to resolve my comment here. I don't have a strong opinion on this anymore and it's probably ok. I don't have a proposal. I just wish the hyperparameters.rst would not mix argparsing and the save_hyperparameters logic, because that one is definitely not legacy.

[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli.rst:123]: That sounds good to me. I can move it in this pull request. How about a "Save hyperparameters" section in https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html?

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst [hunk 6]
```diff
@@ -176,10 +177,13 @@ module and data module.
...
-with LightningCLI is to implement a single module having as init parameters each of the submodules. Since the init
-parameters have as type a class, then in the configuration these would be specified with :code:`class_path` and
-:code:`init_args` entries. For instance a model could be implemented as:
+Many use cases require to have several modules, each with its own configurable options. One possible way to handle this
+with ``LightningCLI`` is to implement a single module having as init parameters each of the submodules. This is known as
+`dependency injection <https://en.wikipedia.org/wiki/Dependency_injection>`__ which is a good approach to improve
+decoupling in your code base.
+
+Since the init parameters of the model have as type hint a class, in the configuration these would be specified with
+``class_path`` and ``init_args`` entries. For instance a model could be implemented as:
```
Comment (rohitgr7) [inline docs/source-pytorch/cli/lightning_cli_advanced_3.rst:186]:
```suggestion
Many use cases require to have several modules, each with its own configurable options. One possible way to handle this
with ``LightningCLI`` is to implement a single module having as init parameters each of the submodules. This is known as
`dependency injection <https://en.wikipedia.org/wiki/Dependency_injection>`__ which is a good approach to improve
decoupling in your code base.

Since the init parameters of the model have as a type hint a class, in the configuration, these would be specified with
``class_path`` and ``init_args`` entries. For instance, a model could be implemented as:
```

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst [hunk 7]
```diff
@@ -209,69 +213,15 @@ If the CLI is implemented as :code:`LightningCLI(MyMainModel)` the configuration
...
 
 Optimizers
 ^^^^^^^^^^
 
-If you will not be changing the class, you can manually add the arguments for specific optimizers and/or
-learning rate schedulers by subclassing the CLI. This has the advantage of providing the proper help message for those
-classes. The following code snippet shows how to implement it:
+In some case it might be desired to fix the optimizer and/or learning scheduler scheduler instead of allowing multiple.
+For this you can manually add the arguments for specific classes by subclassing the CLI. The following code snippet
+shows how to implement it:
```
Comment (rohitgr7) [inline docs/source-pytorch/cli/lightning_cli_advanced_3.rst:224]:
```suggestion
In some cases, fixing the optimizer and/or learning scheduler might be desired instead of allowing multiple.
For this, you can manually add the arguments for specific classes by subclassing the CLI. The following code snippet
shows how to implement it:
```

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst [hunk 8]
```diff
@@ -295,14 +244,14 @@ Therefore, the config file would be structured like:
     trainer:
       ...
 
-Where the arguments can be passed directly through command line without specifying the class. For example:
+where the arguments can be passed directly through command line without specifying the class. For example:
```
Comment (rohitgr7) [inline docs/source-pytorch/cli/lightning_cli_advanced_3.rst:247]:
```suggestion
where the arguments can be passed directly through the command line without specifying the class. For example:
```

File: docs/source-pytorch/cli/lightning_cli_advanced_3.rst [hunk 9]
```diff
@@ -329,11 +278,10 @@ example can be when one wants to add support for multiple optimizers:
...
 
-The value given to :code:`optimizer*_init` will always be a dictionary including :code:`class_path` and
-:code:`init_args` entries. The function :func:`~pytorch_lightning.cli.instantiate_class`
-takes care of importing the class defined in :code:`class_path` and instantiating it using some positional arguments,
-in this case :code:`self.parameters()`, and the :code:`init_args`.
-Any number of optimizers and learning rate schedulers can be added when using :code:`link_to`.
+The value given to ``optimizer*_init`` will always be a dictionary including ``class_path`` and ``init_args`` entries.
+The function :func:`~pytorch_lightning.cli.instantiate_class` takes care of importing the class defined in
+``class_path`` and instantiating it using some positional arguments, in this case ``self.parameters()``, and the
+``init_args``. Any number of optimizers and learning rate schedulers can be added when using ``link_to``.
```
Comment (rohitgr7) [inline docs/source-pytorch/cli/lightning_cli_advanced_3.rst:284]:
```suggestion
The value given to ``optimizer*_init`` will always be a dictionary including ``class_path`` and ``init_args`` entries.
The function :func:`~pytorch_lightning.cli.instantiate_class` takes care of importing the class defined in
``class_path`` and instantiating it using some positional arguments, in this case ``self.parameters()``, and the
``init_args``. Any number of optimizers and learning rate schedulers can be added when using ``link_to``.
```

**Review by rohitgr7 — APPROVED** [review]
great work!
is there a book on lightning cli? 😄

File: docs/source-pytorch/cli/lightning_cli.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (mauvilsa) [reply docs/source-pytorch/cli/lightning_cli.rst:123]: Moved the save_hyperparameters docs to the lightning_module page.

**Comment by mauvilsa** [comment]
Question. Why is this v1.9? It only fixes the documentation of what is currently supported. Nothing related to new features.

**Comment by awaelchli** [comment]
@mauvilsa We don't usually cherry pick docs updates into our bugfix releases (1.8.x), which is why this is labelled 1.9. Once this is merged, the changes will appear [on the website under "latest"](https://pytorch-lightning.readthedocs.io/en/latest/). I think we should move forward and merge this as soon as possible :) 

Note: Might also be worth checking #15473 for possible collisions as it will probably land very soon too.

File: docs/source-pytorch/cli/lightning_cli.rst (hunk 1 above)
Comment (awaelchli) [reply docs/source-pytorch/cli/lightning_cli.rst:6]: What do other reviewers think? @carmocca @rohitgr7 @Borda @RobertLaurella

Comment (Borda) [reply docs/source-pytorch/cli/lightning_cli.rst:6]: cc: @lantiga
